module/VuFind/src/VuFind/Db/Service/TagService.php

```diff
@@ -778,4 +779,401 @@ public function getStatistics(bool $extended = false): array
+     * @param string $text  Tag to look up.
+     *
+     * @return array Tag details.
+     */
+    public function getTagList($sort = 'alphabetical', $limit = 100, $where = null, $text = null)
+    {
+        $tagClause = $this->caseSensitive ? 't.tag' : 'LOWER(t.tag)';
+        $dql = 'SELECT t.id as id, COUNT(DISTINCT(rt.resource)) as cnt, MAX(rt.posted) as posted '
+            . $tagClause . ' AS tag '
+            . 'FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'JOIN rt.tag t ';
+        $parameters = [];
+        switch ($where) {
+            case 'matchText':
+                $dql .= 'WHERE LOWER(t.tag) LIKE :text AND rt.resource is NOT NULL ';
+                $parameters['text'] = $text . '%';
+                break;
+            case 'discardUserList':
+                $dql .= 'WHERE rt.resource is NOT NULL ';
+                break;
+        }
+        $dql .= 'GROUP BY t.id, t.tag ';
+        switch ($sort) {
+            case 'alphabetical':
+                $dql .= 'ORDER BY lower(t.tag), cnt DESC ';
+                break;
+            case 'popularity':
+                $dql .= 'ORDER BY cnt DESC, lower(t.tag) ';
+                break;
+            case 'recent':
+                $dql .= 'ORDER BY posted DESC, cnt DESC, lower(t.tag) ';
+                break;
+        }
+
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $query->setMaxResults($limit);
+        $results = $query->getResult();
+
+        $tagList = [];
+        foreach ($results as $result) {
+            $tagList[] = [
+                'tag' => $result['tag'],
+                'cnt' => $result['cnt'],
+            ];
+        }
+        return $tagList;
+    }
+
+    /**
+     * Get all resources associated with the provided tag query.
+     *
+     * @param string $q      Search query
+     * @param string $source Record source (optional limiter)
+     * @param string $sort   Resource field to sort on (optional)
+     * @param int    $offset Offset for results
+     * @param int    $limit  Limit for results (null for none)
+     * @param bool   $fuzzy  Are we doing an exact or fuzzy search?
+     *
+     * @return array
+     */
+    public function resourceSearch(
+        $q,
+        $source = null,
+        $sort = null,
+        $offset = 0,
+        $limit = null,
+        $fuzzy = true
+    ) {
+        $dql = 'SELECT DISTINCT(r.id) AS resource, r '
+            . 'FROM ' . $this->getEntityClass(Tags::class) . ' t '
+            . 'JOIN ' . $this->getEntityClass(ResourceTags::class) . ' rt WITH t.id = rt.tag '
+            . 'JOIN ' . $this->getEntityClass(Resource::class) . ' r WITH r.id = rt.resource '
+            . 'WHERE rt.resource IS NOT NULL ';
+        $parameters = [];
+        if ($fuzzy) {
+            $dql .= 'AND LOWER(t.tag) LIKE LOWER(:q) ';
+            $parameters['q'] = $q . '%';
```

Comment:
I don't think the wildcard was added in the original code -- it just used `like` instead of `=` for looser matching. Am I overlooking something? If not, I don't think we should change the behavior here.

Reply:
(And if we end up setting the q parameter the same across all cases, we could just move it out of the if statement and initialize `$parameters = compact('q');` above).